File: webapps/frontend/app/viewmodels/containers/failuredomain.js
// Copyright (C) 2016 iNuron NV
//
// This file is part of Open vStorage Open Source Edition (OSE),
// as available from
//
//      http://www.openvstorage.org and
//      http://www.openvstorage.com.
//
// This file is free software; you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
// as published by the Free Software Foundation, in version 3 as it comes
// in the LICENSE.txt file of the Open vStorage OSE distribution.
//
// Open vStorage is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY of any kind.
/*global define */
define([
    'jquery', 'knockout',
    'ovs/generic', 'ovs/api'
], function($, ko, generic, api) {
    "use strict";
    return function(guid) {
        var self = this;

        // Handles
        self.loadHandle = undefined;

        // Observables
        self.address          = ko.observable('');
        self.city             = ko.observable('');
        self.country          = ko.observable('');
        self.disabled         = ko.observable(false);
        self.edit             = ko.observable(false);
        self.guid             = ko.observable(guid);
        self.loaded           = ko.observable(false);
        self.loading          = ko.observable(false);
        self.name             = ko.observable('');
        self.primarySRGuids   = ko.observableArray([]);
        self.secondarySRGuids = ko.observableArray([]);

        // Computed
        self.canSave = ko.computed(function() {
            return !(self.name() === undefined || self.name() === '');
        });
        self.canDelete = ko.computed(function() {
            return self.primarySRGuids().length === 0 && self.secondarySRGuids().length === 0;
        });

        // Functions
        self.fillData = function(data) {
            self.name(data.name);
            generic.trySet(self.city, data, 'city');
            generic.trySet(self.address, data, 'address');
            generic.trySet(self.country, data, 'country');
            if (data.hasOwnProperty('primary_storagerouters_guids')) {
                self.primarySRGuids(data.primary_storagerouters_guids);
            }
            if (data.hasOwnProperty('secondary_storagerouters_guids')) {
                self.secondarySRGuids(data.secondary_storagerouters_guids);
            }

            self.loaded(true);
            self.loading(false);
        };
        self.load = function() {
            return $.Deferred(function(deferred) {
                self.loading(true);
                if (generic.xhrCompleted(self.loadHandle)) {
                    self.loadHandle = api.get('failure_domains/' + self.guid())
                        .done(function(data) {
                            self.fillData(data);
                            deferred.resolve();
                        })
                        .fail(deferred.reject)
                        .always(function() {
                            self.loading(false);
                        });
                } else {
                    deferred.reject();
                }
            }).promise();
        };
        self.save = function() {
            return $.Deferred(function(deferred) {
                var data = {
                    name: self.name(),
                    address: self.address(),
                    city: self.city(),
                    country: self.country()
                };
                if (self.guid() === undefined) {
                    api.post('failure_domains', { data: data })
                        .done(function (data) {
                            deferred.resolve(data.guid);
                        })
                        .fail(deferred.reject)
                        .always(function() {
                            self.edit(false);
                        });
                } else {
                    api.patch('failure_domains/' + self.guid(), { data: data })
                        .done(function (data) {
                            deferred.resolve(data.guid);
                        })
                        .fail(deferred.reject)
                        .always(function() {
                            self.edit(false);
                        });
                }
            }).promise();
        };
    };
});

File: ovs/dal/hybrids/failuredomain.py
# Copyright (C) 2016 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.

"""
FailureDomain module
"""
from ovs.dal.dataobject import DataObject
from ovs.dal.structures import Property


class FailureDomain(DataObject):
    """
    The FailureDomain class represents a domain where DTL and/or volumedriver MDS services can be configured to
     Eg:
     || Storage Router || Failure Domain || Backup Failure Domain ||
     |        sr1       |       fd1       |          fd2           |
     |        sr2       |       fd1       |          fd2           |
     |        sr3       |       fd2       |          fd1           |
     |        sr4       |       fd2       |          fd1           |
      - Storage router 1 is part of failure domain 1 and backup failure domain 2
      - Storage router 2 is part of failure domain 1 and backup failure domain 2
      - Storage router 3 is part of failure domain 2 and backup failure domain 1
      - Storage router 4 is part of failure domain 2 and backup failure domain 1

      Storage router 1 will have its DTL configured within its own failure domain (by default), which means on storage router 2

      Each storage router CAN also have a backup failure domain
      For storage router 1 this will be backup failure domain 2 and for storage router 4 this will be backup failure domain 1
      If storage router 2 would go down, the backup failure domain will be used, which means storage router 3 and storage router 4 can be used for DTL of storage router 1
    """
    __properties = [Property('name', str, doc='The name for the (backup) failure domain'),
                    Property('address', str, mandatory=False, doc='Address of the domain'),
                    Property('city', str, mandatory=False, doc='City where domain is located'),
                    Property('country', str, mandatory=False, doc='Country where domain is located')]
    __relations = []
    __dynamics = []
